samples/net/prometheus/src/main.c

```diff
@@ -0,0 +1,240 @@
+#include <zephyr/net/socket.h>
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/counter.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_DBG);
+
+struct {
+
+	struct prometheus_collector *collector;
+
+	struct prometheus_counter *counter;
+
+} prom_context;
+
+int sendall(struct http_client_ctx *client, const void *buf, size_t len)
+{
+	while (len) {
+		ssize_t out_len = zsock_send(client->fd, buf, len, 0);
+
+		if (out_len < 0) {
+			return -errno;
+		}
+
+		buf = (const char *)buf + out_len;
+		len -= out_len;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_NET_SAMPLE_HTTP_SERVICE)
+static uint16_t test_http_service_port = CONFIG_NET_SAMPLE_HTTP_SERVER_SERVICE_PORT;
+HTTP_SERVICE_DEFINE(test_http_service, CONFIG_NET_CONFIG_MY_IPV4_ADDR, &test_http_service_port, 1,
+		    10, NULL);
+
+static uint8_t recv_buffer[2048];
+
+static int dyn_handler(struct http_client_ctx *client, enum http_data_status status,
+		       uint8_t *buffer, size_t len, void *user_data)
+{
+#define MAX_TEMP_PRINT_LEN 32
+	int rc;
+	char hex_len[MAX_TEMP_PRINT_LEN];
+	static char print_str[MAX_TEMP_PRINT_LEN];
+	enum http_method method = client->method;
+	static size_t processed;
+
+	__ASSERT_NO_MSG(buffer != NULL);
+
+	if (status == HTTP_SERVER_DATA_ABORTED) {
+		LOG_DBG("Transaction aborted after %zd bytes.", processed);
+		processed = 0;
+		return 0;
+	}
+
+	processed += len;
+
+	snprintf(print_str, sizeof(print_str), "%s received (%zd bytes)", http_method_str(method),
+		 len);
+	LOG_HEXDUMP_DBG(buffer, len, print_str);
+
+	if (status == HTTP_SERVER_DATA_FINAL) {
+		LOG_DBG("All data received (%zd bytes).", processed);
+		processed = 0;
+	}
+
+	/* incrase counter per request */
+	prometheus_counter_inc(prom_context.counter);
+
+	(void)memset(recv_buffer, 0, sizeof(recv_buffer));
+
+	prometheus_format_exposition(prom_context.collector, recv_buffer, sizeof(recv_buffer));
+
+	/**
+	 * The response status line indicates that the request was successful (200 OK).
+	 * The response headers include Date, Server, Content-Type, and Transfer-Encoding.
+	 * Transfer-Encoding is set to chunked, indicating that the response body is sent in a
+	 * series of chunks. Each chunk begins with the hexadecimal length of the chunk, followed by
+	 * a carriage return and line feed ("\r\n"), then the chunk data itself, and finally
+	 * terminated by another carriage return and line feed ("\r\n"). The last chunk has a length
+	 * of 0, indicating the end of the response body.
+	 *
+	 */
+
+	rc = snprintk(hex_len, sizeof(hex_len), "%x\r\n", strlen(recv_buffer));
+	if (rc < 0) {
+		LOG_ERR("Cannot format data length (%d)", rc);
+		return rc;
+	}
+
+	rc = sendall(client, hex_len, rc);
```

Comment:
I'm a bit puzzled here, why are we sending the reply back to the client directly from the resource handler? Wouldn't that be in conflict with the server code, which actually is responsible for sending the reply back to the client?

I would expect, that the application callback simply fills the `buffer` with the response payload, and return the number of bytes written back to the server. The server should then send them back in a HTTP "chunk".

Or am I missing something here? For reference:
https://docs.zephyrproject.org/latest/connectivity/networking/api/http_server.html#dynamic-resources

Reply:
I'm totally aggree with you. If review the `dynamic_get_req` callback you will see that part:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/3eec4384-86b1-4c3a-b2b7-2dd4241fd00c)

When I try to copy `dynamic_get_req` buffer callback hang infinite loop because callback return check one time.

Reply:
The callback needs to return 0 to break the loop, that tells the server side that no more data is expected to be included in the response. So in a simple case, I imagine callback gets called twice, the first time you fill out the buffer and return number of bytes written, the second just return 0 indicating that no more data is to be expected.

Reply:
I did try first is like your tell and noticed to written size not passing to callback so cant check how many bytes are sent.

Reply:
The server guarantees to send all bytes that have been returned by the callback:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/http/http_server_http1.c#L150

Reply:
Yes, you'r right. I try to say the server pass `copy_len` argument to callback. The server calculate `copy_len` size zero every time: https://github.com/zephyrproject-rtos/zephyr/blob/62893328293f8719f72b0a4e2edcb48fef3e125c/samples/net/sockets/http_server/src/main.c#L61

Our scenario following steps:

- Call `dynamic_get_req` with parameters `copy_len` is 0.
- We'r copy our buffer to the server buffer and return buffer size.
- The server perform callback again because `send_len` is bigger than 0.
- We give `copy_len`  argument as zero and copy buffer again.

How do we check that situation?

Reply:
I think I figured it out :). CI issues and rebase done.